components/sup/src/manager/file_watcher.rs

```diff
@@ -1,3464 +1,2652 @@
-        debug!("add leaf file: {}", dits!(watched_file));
-        let dir_file_name = watched_file.get_common().dir_file_name.clone();
-        let needs_watch = self.add_dir(&dir_file_name);
-        let chain_link_info = match self.paths.entry(dir_file_name.as_path()) {
-            Entry::Occupied(o) => {
-                error!("Paths inconsistency in add_leaf_watched_file, expect strange results");
-                o.get().get_common().get_chain_link_info()
-            }
-            Entry::Vacant(v) => v.insert(watched_file).get_common().get_chain_link_info(),
-        };
-        if needs_watch {
-            LeafResult::NewInNewDirectory(chain_link_info, dir_file_name.directory)
-        } else {
-            LeafResult::NewInOldDirectory(chain_link_info)
-        }
-    }
+                assert_eq!(1, fw.callbacks.callback_type_count(Appeared));
+                assert_eq!(1, fw.callbacks.path_instance_count(&rtd_watched_file));
 
-    fn get_or_add_directory(&mut self, common: Common) -> BranchResult {
-        self.get_or_add_branch_watched_file(WatchedFile::Directory(common))
-    }
+                Ok(())
+            }
 
-    fn get_or_add_symlink(&mut self, common: Common) -> BranchResult {
-        self.get_or_add_branch_watched_file(WatchedFile::Symlink(common))
-    }
+            #[test]
+            fn create_watched_file_in_ancestral_directory_pw() -> Result<()> {
+                let lock = lock_env_var();
+                lock.set("aarch64-darwin");
+                let r = create_watched_file_in_ancestral_directory();
+                lock.unset();
+                r
+            }
 
-    fn get_or_add_branch_watched_file(&mut self, watched_file: WatchedFile) -> BranchResult {
-        debug!("get or add branch file: {}", dits!(watched_file));
-        if self.paths.contains_key(&watched_file.get_common().path) {
-            return BranchResult::AlreadyExists;
-        }
+            #[test]
+            fn create_watched_file_in_descendant_directory() -> Result<()> {
+                let cur_dir = PathBuf::from(".");
+                let descendent_dir = cur_dir.join(fs::make_uniq("dir"));
+                let descendent_watched_file = descendent_dir.join(fs::make_uniq("file"));
+                let tc = TestCallbacks::default();
+                let mut fw = Tfw::new(&descendent_watched_file, tc, true)?;
 
-        let dir_file_name = watched_file.get_common().dir_file_name.clone();
+                // we're watching the relative path but creating the file with the absolute path
+                fs::create_dir(&descendent_dir)?;
+                fs::create_file(&descendent_watched_file)?;
 
-        let needs_watch = self.add_dir(&dir_file_name);
-        let chain_link_info = watched_file.get_common().get_chain_link_info();
-        debug!("chain link info: {:?}", dits!(chain_link_info));
-        self.paths.insert(dir_file_name.as_path(), watched_file);
+                for _ in 0..max_test_attempts(&fw) {
```

Comment:
do you need this loop for creation scenarios?

Reply:
Pretty sure I need the loop.  Once the thing that triggers the event happens, here fs:create_file call, the code has to receive the event which happens in single_iteration.